### ISSUE TITLE
Centrally installed hooks vs. build reproducibility

### ISSUE THREAD

**Issue by sztomi** [conan-io/hooks#35]
Continuing discussion from https://github.com/conan-io/hooks/pull/21#issuecomment-461097566

In my opinion, hooks, as they are currently, should be akin to git hooks: not necessary for functioning, but allow complementing functionality. They should never affect the resulting build.

If I relied on a hook for capturing some build-time information in my package (as suggested on Slack, see [my earlier comment](https://github.com/conan-io/hooks/pull/21#issuecomment-460703534), I would need a specific hook installed to be able to reproduce that package. And that requirement is not apparent from the conanfile (nor is it enforced by conan). My recipe is no longer portable, I need to take specific steps (installing a hook) before I can rebuild it. 

Furthermore, it seems to me that once a hook is installed it affects all builds? (I might be wrong about this). This would make it impossible to perform functionality on just a couple of recipes. 

I would propose a different, simpler model, based on function overrides. So hook functions like `pre_export`, `post_export` etc. would be simply member functions in Conanfile. That would make using the hooks very visible and environment-independent (if I clone the repo on a different machine, I don't need to perform additional hidden steps, it's there in the conanfile). This model also allows global overrides, by relying on `python_requires` and an intermediary base class:

```py
class MyConanfile(Conanfile):
  def post_source(self):
    print("I'm in all recipes derived from MyConanfile")
```

Again, this preserves the build reproducibility.

**Comment by jgsogo** [comment]
Yes, you are right, we are aware of that concern too. We do not want hooks to create a _fork_ of Conan, or modify its behavior, but there are some functionalities that I wouldn't know how to implement otherwise.

The aim of the hooks is to add custom behavior, common for all the recipes, but specific for an user or company; for example, the _export-metadata_ hook is adding information to the package without modifying the recipes, and this information could be required by the CI system.

Other hooks we are proposing are just checking that the recipes are correct, checking some aspects of the generated binaries, updating data in Bintray or Github,... but we are not trying to change anything in the package itself.

I agree that the problem exists because we cannot control the hooks people will implement and use on its own, that risk will always be there and given the flexibility of Python it will be impossible to control.



Your approach using functions and inheritance with _python_requires_ is very interesting and fully reproducible, but it requires modifying the recipe to change or add some behavior. Also, in order to run a recipe, the corresponding _python_requires_ are needed and it can discourage companies to adopt existing recipes or sharing their own.

Maybe, the best protection against hooks that modifies the Conan logic is that those hooks will be applied to every recipe, I don't expect shared hooks to contain `if` clauses that apply only to a specific recipe (`if reference == "name/ver@user/channel": ...`).

This is just my opinion, I think hooks are needed and I like the way they work right now. But let's hear what others have to say.

Thanks!

**Comment by sztomi** [comment]
> it requires modifying the recipe to change or add some behavior

That's precisely the point: otherwise this is an invisible change. I have nothing against the hooks that check correctness and such, as they don't affect the build output directly. I'd argue that adding metadata does though.

> the corresponding _python_requires_ are needed and it can discourage companies to adopt existing recipes or sharing their own.

Well it's not much different from an extra requires line, is it? Also, I think the idea that companies will start to share their internal recipes is a bit unrealistic, so optimizing for that scenario is a wasted effort. Not because they (we) are against sharing, but because the recipes we build are tied to our infrastructure and not easily portable for the outside world. As such, it requires disproportional effort, especially if we were to not only contribute but maintain recipes. I think the split between corporate and OSS use will continue to exist, and that is fine.

> Maybe, the best protection against hooks that modifies the Conan logic is that those hooks will be applied to every recipe

That sounds like a compromise when there is no need for one.

**Comment by jgsogo** [comment]
I agree with you. I don't like the idea of generating different outputs with the same recipe (I won't use an environment variable in a recipe), it is something we have to avoid. The `export_metadata` hook is like a proof of concept (related to the [`artifacts.properties` file](https://docs.conan.io/en/latest/reference/config_files/artifacts.properties.html)) where a user needs a different way to store/access metadata related to the packages. Maybe we shouldn't ever merge it into this repo and leave it like an example in a blogpost related to CI flows.

**Comment by jgsogo** [comment]
Hi @sztomi! @danimtb is writing a blog post about hooks that will be published soon, I would want your concerns (https://github.com/conan-io/hooks/issues/35) to be taking into account there. I think there is already a warning about it, but if you think we should remark it a little bit more, we will be pleased to receive your feedback there too.

Thanks!

**Comment by sztomi** [comment]
@jgsogo sorry I totally missed this notification! I read the post and I think it's fine :)
